internal/provider/kubernetes/controller.go

```diff
@@ -180,213 +181,206 @@ func (r *gatewayAPIReconciler) Reconcile(ctx context.Context, _ reconcile.Reques
-			if err := r.gatewayClassUpdater(ctx, acceptedGC, false, string(gwapiv1.GatewayClassReasonInvalidParameters), msg); err != nil {
-				r.log.Error(err, "unable to update GatewayClass status")
-			}
-			r.log.Error(err, "failed to process parametersRef for gatewayclass", "name", acceptedGC.Name)
+		if resourcesMap[acceptedGC.Name].EnvoyProxy != nil && resourcesMap[acceptedGC.Name].EnvoyProxy.Spec.MergeGateways != nil {
+			r.mergeGateways[acceptedGC.Name] = *resourcesMap[acceptedGC.Name].EnvoyProxy.Spec.MergeGateways
+		}
+		if err := r.gatewayClassUpdater(ctx, acceptedGC, true, string(gwapiv1.GatewayClassReasonAccepted), status.MsgValidGatewayClass); err != nil {
+			r.log.Error(err, "unable to update GatewayClass status")
 			return reconcile.Result{}, err
 		}
-	}
-
-	if resourceTree.EnvoyProxy != nil && resourceTree.EnvoyProxy.Spec.MergeGateways != nil {
-		r.mergeGateways = *resourceTree.EnvoyProxy.Spec.MergeGateways
-	}
-
-	if err := r.gatewayClassUpdater(ctx, acceptedGC, true, string(gwapiv1.GatewayClassReasonAccepted), status.MsgValidGatewayClass); err != nil {
-		r.log.Error(err, "unable to update GatewayClass status")
-		return reconcile.Result{}, err
-	}
 
-	// Update finalizer on the gateway class based on the resource tree.
-	if len(resourceTree.Gateways) == 0 {
-		r.log.Info("No gateways found for accepted gatewayclass")
+		// Update finalizer on the gateway class based on the resource tree.
+		if len(resourcesMap[acceptedGC.Name].Gateways) == 0 {
+			r.log.Info("No gateways found for accepted gatewayclass")
 
-		// If needed, remove the finalizer from the accepted GatewayClass.
-		if err := r.removeFinalizer(ctx, acceptedGC); err != nil {
-			r.log.Error(err, fmt.Sprintf("failed to remove finalizer from gatewayclass %s",
-				acceptedGC.Name))
-			return reconcile.Result{}, err
-		}
-	} else {
-		// finalize the accepted GatewayClass.
-		if err := r.addFinalizer(ctx, acceptedGC); err != nil {
-			r.log.Error(err, fmt.Sprintf("failed adding finalizer to gatewayclass %s",
-				acceptedGC.Name))
-			return reconcile.Result{}, err
+			// If needed, remove the finalizer from the accepted GatewayClass.
+			if err := r.removeFinalizer(ctx, acceptedGC); err != nil {
+				r.log.Error(err, fmt.Sprintf("failed to remove finalizer from gatewayclass %s",
+					acceptedGC.Name))
+				return reconcile.Result{}, err
+			}
+		} else {
+			// finalize the accepted GatewayClass.
+			if err := r.addFinalizer(ctx, acceptedGC); err != nil {
+				r.log.Error(err, fmt.Sprintf("failed adding finalizer to gatewayclass %s",
+					acceptedGC.Name))
+				return reconcile.Result{}, err
+			}
 		}
+		// The Store is triggered even when there are no Gateways associated to the
+		// GatewayClass. This would happen in case the last Gateway is removed and the
+		// Store will be required to trigger a cleanup of envoy infra resources.
+		r.resources.GatewayAPIResources.Store(acceptedGC.Name, resourcesMap[acceptedGC.Name])
```

Comment:
seeing a bug 
* the gateway-api translator works at a few gateway class level 
https://github.com/envoyproxy/gateway/blob/e168ef961dbd40861af401bf70c781e2cba282fd/internal/gatewayapi/runner/runner.go#L63
* however in the gateway-api layer, the logic to clean up IRs is global, which will cause other gateways to be deleted
https://github.com/envoyproxy/gateway/blob/e168ef961dbd40861af401bf70c781e2cba282fd/internal/gatewayapi/runner/runner.go#L172

Possible solution
* filter on relevant keys (try and match GC) when cleaning
* or keep it flat, and pass 1 provider resource to the gateway-api layer with a list of GCs (the gateway-api translator can still work on 1 GC at a time)

Reply:
right I checked it out, thanks @arkodg